### PR TITLE
feat: Forms: Add reset validation functionality to form and form element [GAUD-6979]

### PR DESCRIPTION
**Pull request by margaree**
[Jira ticket](https://desire2learn.atlassian.net/browse/GAUD-6979)

This adds reset functionality to both a `d2l-form` and `FormElementMixin`. I'd originally looked at adding it to the `FormMixin` but that got complicated because of potential for nesting of forms and how some functionality lives in Form vs others in `FormMixin`. I can re-look at that, but at the moment there's nothing just using `FormMixin`.

This can be tested on the form demo page.

To do:
- [x] tests
- [x] documentation
- [x] Add flag

### CODE REVIEW

File: components/form/form-element-mixin.js [hunk 1]
```diff
@@ -227,6 +227,18 @@ export const FormElementMixin = superclass => class extends LocalizeCoreElement(
 		await this.updatedComplete;
 	}
 
+	reset() {
+		this.invalid = false;
+		this.validationError = null;
+		this._errors = [];
+
+		this.childErrors.forEach((_, ele) => {
+			if (!isCustomFormElement(ele)) return;
```
Comment (margaree) [inline components/form/form-element-mixin.js:236]: A case (?) that wouldn't be covered is if the child error is a native element. I don't know if that's actually a thing we deal with.

File: components/form/form-element-mixin.js [hunk 2]
```diff
@@ -227,6 +227,18 @@ export const FormElementMixin = superclass => class extends LocalizeCoreElement(
 		await this.updatedComplete;
 	}
 
+	reset() {
+		this.invalid = false;
+		this.validationError = null;
+		this._errors = [];
```
Comment (margaree) [inline components/form/form-element-mixin.js:233]: Let me know if there's anything else we should be resetting here.

Comment (dlockhart) [reply components/form/form-element-mixin.js:233]: Does this reset the element _value_ as well, or just its error state? When I think about what "resetting a form" would do, I think I'd expect it to act like `<button type="reset">` where it also resets values to their "initial" state.

I'm not sure whether the case we're solving for requires a full reset, but if not maybe we should call this "resetValidation"?

Comment (margaree) [reply components/form/form-element-mixin.js:233]: It wouldn't reset the values. Agreed we're just solving for a validation reset so I'll rename.

File: components/form/form.js [hunk 3]
```diff
@@ -77,6 +77,26 @@ class Form extends FormMixin(LitElement) {
 		this._submitData(submitter);
 	}
 
+	resetValidation() {
```
Comment (dlockhart) [inline components/form/form.js:80]: Up for debate: should forms listen for the `d2l-dialog-close` event and automatically reset validation? That would solve the consumer case that came up (and presumably future similar cases), but obviously there's a risk that there are forms inside dialogs that wouldn't want this. We could flag that portion of the change to help mitigate that. Just trying to avoid everyone needing to call this when dialog's close if we could handle it for them.

Comment (dbatiste) [reply components/form/form.js:80]: I think I could get onboard with that. That's probably the behaviour I would want, and if the edge-case arose, we could provide a way to opt out of that.

Comment (margaree) [reply components/form/form.js:80]: I like that idea. Will update

File: components/form/test/form-element.test.js [hunk 4]
```diff
@@ -196,6 +196,16 @@ describe('form-element', () => {
...
 
+	describe('resetValidation', () => {
+		it('should show reset validation errors', async() => {
+			await formElement.requestValidate();
+			expect(formElement.validationError).to.equal('Test form element is required.');
+			formElement.resetValidation();
+			expect(formElement.invalid).to.be.false;
+			expect(formElement.validationError).to.be.null;
+		});
+	})
```
Comment (dlockhart) [inline components/form/test/form-element.test.js]:
```suggestion
	});
```

**Review by dlockhart — APPROVED** [review]

File: components/form/form.js [hunk 5]
```diff
@@ -58,7 +59,13 @@ class Form extends FormMixin(LitElement) {
 	firstUpdated(changedProperties) {
 		super.firstUpdated(changedProperties);
 
-		this.addEventListener('d2l-dialog-close', () => {
+		const dialogAncestor = findComposedAncestor(
+			this,
+			(node) => { return (node?.tagName?.includes('D2L-DIALOG')); }
```
[the remark this answers is not in the document]
Comment (dlockhart) [reply components/form/form.js:64]: The `.includes(` should cover that?

[the remark this answers is not in the document]
Comment (dbatiste) [reply components/form/form.js:64]: Oh good point. I glossed over that. So it would just be anyone (like LCS) that might have extended the mixin for their own component. Maybe they're on their own to reset their forms.

[the remark this answers is not in the document]
Comment (dlockhart) [reply components/form/form.js:64]: Wonder if something like `this.prototype instanceof DialogMixin` would work?

[the remark this answers is not in the document]
Comment (margaree) [reply components/form/form.js:64]: Still looking into it but so far haven't been able to get that to work

[the remark this answers is not in the document]
Comment (dbatiste) [reply components/form/form.js:64]: I know in a couple other places we've set a property that can be checked, which seems kinda hacky. What Dave proposed seems much nicer.

[the remark this answers is not in the document]
Comment (margaree) [reply components/form/form.js:64]: I haven't been able to get anything along the lines of the `node.prototype instanceof DialogMixin` check working. I can get setting a property on dialog mixin and checking for that working. Do we want to go through with the property option? Any other ideas that I maybe haven't tried?

[the remark this answers is not in the document]
Comment (dlockhart) [reply components/form/form.js:64]: Hmm, I feel like a `_isDialogMixin = true` "private" variable would be slightly more resilient than the tag name check, but am good with either.

**Review by dlockhart — APPROVED** [review]
